src/expr/src/relation/mod.rs

```diff
@@ -1739,11 +1740,42 @@ impl MirRelationExpr {
         false
     }
 
+    /// Given the ids and values of a LetRec, it computes the subset of ids that are used across
+    /// iterations.
+    pub fn ids_used_across_iterations(
+        ids: &Vec<LocalId>,
+        values: &Vec<MirRelationExpr>,
+    ) -> Result<BTreeSet<LocalId>, RecursionLimitError> {
+        let mut used_across_iterations = BTreeSet::new();
+        let mut defined = BTreeSet::new();
+        for (binding_id, value) in ids.iter().zip(values.iter()) {
```

Comment:
nit: use `itertools::zip_eq(ids.iter(), values.iter())` so we panic if for some reason the two vectors differ in their size.

Reply:
Neat, done!

Reply:
Btw. there is also a `zip` call that I introduced in `predicate_pushdown.rs`, and I tried to replace that too, but that one is not possible to replace because of the `rev()`: "the trait `DoubleEndedIterator` is not implemented for `ZipEq`"

Reply:
That is fine. I just saw that used in some other part of the code so I thought it's a safer bet to catch invalid states early, so I mentioned it, but we don't have to go above and beyond to introduce it everywhere.